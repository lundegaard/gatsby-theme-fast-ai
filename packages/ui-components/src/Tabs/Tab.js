import React, { forwardRef, useContext } from 'react';
import PropTypes from 'prop-types';

import Flex from '../Flex';
import { useAbstractButton } from '../AbstractButton';

import { TabListContext } from './contexts';

const styles = {
	root: {
		alignItems: 'stretch',
		variant: 'Tab',
		flexShrink: 0,
		flexGrow: 1,
		flexBasis: 'auto',
		flexDirection: 'column',
		minWidth: '0px',
		minHeight: '0px',
		position: 'relative',
		zIndex: 0,
		borderBottom: theme => theme.separatorBorder.bold,
	},
	label: {
		variant: 'TabLabel',
		cursor: 'pointer',
		flexDirection: 'column',
		flexGrow: 1,
		flexShrink: 0,
		flexBasis: 'auto',
		justifyContent: 'center',
		outlineStyle: 'none',
		alignItems: 'center',
		minWidth: '0px',
		minHeight: '0px',
		position: 'relative',
		zIndex: 0,
	},
};

const Tab = forwardRef(({ sx, label, value, ...rest }, ref) => {
	const { dispatch, state: currentValue } = useContext(TabListContext);

	const buttonProps = useAbstractButton({
		onClick: () => dispatch({ type: 'change', payload: value }),
	});

	return (
		<Flex
			sx={{
				...sx,
				...styles.root,
				borderBottomColor: theme =>
					currentValue === value ? theme.colors.secondary : 'transparent',
			}}
			{...rest}
			{...buttonProps}
			ref={ref}
		>
			<Flex sx={styles.label}>{label}</Flex>
		</Flex>
	);
});

Tab.propTypes = {
	label: PropTypes.node,
	value: PropTypes.any,
};

Tab.displayName = 'Tab';

export default Tab;
