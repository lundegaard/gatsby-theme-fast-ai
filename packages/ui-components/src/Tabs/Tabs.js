import React, { forwardRef, useEffect, useMemo, useReducer } from 'react';
import PropTypes from 'prop-types';

import Flex from '../Flex';
import Box from '../Box';

import { TabListContext, TabsContext } from './contexts';

export const TabPane = forwardRef(({ children, ...rest }, ref) => (
	<Box {...rest} ref={ref}>
		{children}
	</Box>
));

TabPane.propTypes = {
	children: PropTypes.node,
};

TabPane.displayName = 'TabPane';

export const TabPanels = forwardRef(({ children, ...rest }, ref) => (
	<Box {...rest} ref={ref}>
		{children({ getAriaProps: () => ({}) })}
	</Box>
));

TabPanels.propTypes = {
	children: PropTypes.node,
};

TabPanels.displayName = 'TabPanels';

const reducer = (state, action) => {
	switch (action.type) {
		case 'change':
			return action.payload;
		default:
			return state;
	}
};

export const TabList = forwardRef(
	({ children, sx, value = '', onChange, ...rest }, ref) => {
		const [state, dispatch] = useReducer(reducer, value);
		const api = useMemo(() => ({ state, dispatch }), [state, dispatch]);

		useEffect(() => {
			onChange?.(state);
		}, [state]);

		useEffect(() => {
			dispatch({ type: 'change', payload: value });
		}, [value]);

		return (
			<TabListContext.Provider value={api}>
				<Box
					ref={ref}
					{...rest}
					sx={{
						width: '100%',
						overflow: 'hidden',
						...sx,
					}}
				>
					<Flex
						role="tablist"
						sx={{
							borderBottom: theme => theme.separatorBorder.normal,
							width: '100%',
							alignItems: 'stretch',
							flexBasis: 'auto',
							overflowX: 'auto',
							overflowY: 'hidden',
							overscrollBehaviorY: 'none',
							scrollPadding: '0',
							scrollSnapType: 'x mandatory',
							'&::-webkit-scrollbar': { display: 'none' },
							msOverflowStyle: 'none',
							scrollbarWidth: 'none',
						}}
					>
						{children}
					</Flex>
				</Box>
			</TabListContext.Provider>
		);
	},
);

TabList.propTypes = {
	children: PropTypes.node,
	onChange: PropTypes.func,
	value: PropTypes.any,
};

TabList.displayName = 'TabList';

export const Tabs = forwardRef(({ children, name, ...rest }, ref) => (
	<TabsContext.Provider value={name}>
		<Box {...rest} ref={ref}>
			{children}
		</Box>
	</TabsContext.Provider>
));

Tabs.propTypes = {
	/** Children to be rendered in the main container. */
	children: PropTypes.node,
	name: PropTypes.string,
};

Tabs.displayName = 'Tabs';

export default Tabs;
