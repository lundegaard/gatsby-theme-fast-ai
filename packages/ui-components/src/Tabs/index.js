export { default } from './Tabs';
export { Tabs, TabList, TabPane, TabPanels } from './Tabs';
export { default as Tab } from './Tab';
