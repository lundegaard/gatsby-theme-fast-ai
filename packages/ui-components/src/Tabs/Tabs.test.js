import React from 'react';
import { shallow } from 'enzyme';

import Tabs from './Tabs';

describe('<Tabs />', () => {
	const requiredProps = {};

	it('matches snapshot when shallowly rendered', () => {
		const wrapper = shallow(<Tabs {...requiredProps} />);
		expect(wrapper).toMatchSnapshot();
	});
});
