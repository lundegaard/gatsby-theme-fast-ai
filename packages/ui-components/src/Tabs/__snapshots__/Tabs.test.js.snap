// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Tabs /> matches snapshot when shallowly rendered 1`] = `
<ContextProvider>
  <Box />
</ContextProvider>
`;
